### PR TITLE
Fix gestures in Label Spans

### LINKED ISSUE #4734: Gestures don't work on Label Spans
### Description

I added a gesture to a label span and tapped it. Nothing is happening on Windows. Cannot test Android bc it doesn't render the span.

Platforms:

- [x] Android https://github.com/dotnet/maui/pull/14410
- [x] iOS https://github.com/dotnet/maui/pull/15544
- [x] Mac Catalyst https://github.com/dotnet/maui/pull/15544
- [ ] Windows https://github.com/dotnet/maui/pull/17731

### Steps to Reproduce

Run this:

```xaml
<?xml version="1.0" encoding="utf-8" ?>
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             x:Class="ControlGallery.Pages.LabelPage"
             Shell.FlyoutBehavior="Disabled"
             >

    <ContentPage.Content>
        <ScrollView>
            <StackLayout Padding="{OnPlatform iOS='30,60,30,30', Default='30'}">

                <Label LineBreakMode="NoWrap" LineHeight="1.4">
                    <Label.FormattedText>
                        <FormattedString>
                            <Span Text="Font name: Default &#10;"/>
                            <Span Text="Version: 1.00  &#10;"/>
                            <Span Text="Digitally Signed, TrueType Outlines &#10;"/>
                            <Span Text="abcdefghijklmnopqrstuvwxyz "/>
                            <Span Text="abcdefghijklmnopqrstuvwxyz &#10;" TextTransform="Uppercase"/>
                            <Span Text="1234567890.:,;'+-*/=  &#10;"/>
                            <Span Text="12 The quick brown fox jumps over the lazy dog. 1234567890 &#10;" FontSize="12"/>
                            <Span Text="18 The quick brown fox jumps over the lazy dog. 1234567890 &#10;" FontSize="18"/>
                            <Span Text="24 The quick brown fox jumps over the lazy dog. 1234567890 &#10;" FontSize="24"/>
                            <Span Text="36 The quick brown fox jumps over the lazy dog. 1234567890 &#10;" FontSize="36"/>
                            <Span Text="48 The quick brown fox jumps over the lazy dog. 1234567890 &#10;" FontSize="48"/>
                            <Span Text="60 The quick brown fox jumps over the lazy dog. 1234567890 &#10;" FontSize="60"/>
                            <Span Text="72 The quick brown fox jumps over the lazy dog. 1234567890 " FontSize="72" >
                                <Span.GestureRecognizers>
                                    <TapGestureRecognizer Command="{Binding TapCommand}" CommandParameter="https://google.com" />
                                </Span.GestureRecognizers>
                            </Span>
                        </FormattedString>
                    </Label.FormattedText>                    
                </Label>
            </StackLayout>
        </ScrollView>

    </ContentPage.Content>
</ContentPage>
```

```csharp
using System.Windows.Input;

namespace ControlGallery.Pages;

public partial class LabelPage : ContentPage
{
    public ICommand TapCommand => new Command<string>(async (url) => await Launcher.OpenAsync(url));

    public LabelPage()
    {
        InitializeComponent();
        BindingContext = this;
    }
}

```

image

Screenshot_1645061133

### Version with bug

Preview 13 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android, Windows

### Affected platform versions

All

### Did you find any workaround?

No workaround

### Relevant log output

_No response_

### LINKED ISSUE #8004: TapGestureRecognizer does not work in label's span
### Description

In the following xaml, tap does't work and react to the finger tap.

            <Label>
                <Label.FormattedText>
                    <FormattedString>
                        <Span Text="Alternatively, click " />
                        <Span Text="here"
                  TextColor="Blue" FontSize="24"
                  TextDecorations="Underline">
                            <Span.GestureRecognizers>
                                <TapGestureRecognizer Command="{Binding TapCommand}"
                                          CommandParameter="https://docs.microsoft.com/dotnet/maui/" />
                            </Span.GestureRecognizers>
                        </Span>
                        <Span Text=" to view .NET MAUI documentation." />
                    </FormattedString>
                </Label.FormattedText>
            </Label>

### Steps to Reproduce

create an maui app
add label and some span
add TapGestureRecognizer 

### Version with bug

6.0 (current)

### Last version that worked well

Unknown/Other

### Affected platforms

Android, Windows

### Affected platform versions

Android 10、windows sdk 10.0.19041.0

### Did you find any workaround?

no workaround. but TapGestureRecognizer  in label works like the following xaml:
            <Label Text="here"
                  TextColor="Blue" FontSize="24"
                  TextDecorations="Underline">
                <Label.GestureRecognizers>
                    <TapGestureRecognizer Command="{Binding TapCommand}"
                                          CommandParameter="https://docs.microsoft.com/dotnet/maui/" />
                </Label.GestureRecognizers>
            </Label>

### Relevant log output

```shell
no
```

### PR DESCRIPTION
**Pull request by jsuarezruiz**
### Description of Change

Fix gestures in Label Spans.

Android
fix-gestures-span-droid
image

iOS/Catalyst
fix-gestures-span-ios

Windows
(Work in progress)

To test/validate the changes, open the .NET MAUI Gallery and navigate to the Label page. Tap the Span with the text "Click Me". If the Span changes the TextColor, is working as expected.

### Issues Fixed

Fixes #4734 
Fixes #8004

### CODE REVIEW

File: src/Controls/src/Core/Platform/Windows/Extensions/FormattedStringExtensions.cs [hunk 1]
```diff
@@ -128,5 +129,54 @@ public static void UpdateInlines(this TextBlock textBlock, Label label)
...
+
+			if (labelWidth <= 0 || control.Height <= 0)
+				return;
+
+			for (int i = 0; i < element.FormattedText.Spans.Count; i++)
+			{
+				var span = element.FormattedText.Spans[i];
+
+				var inline = control.Inlines.ElementAt(i);
+				var startRect = inline.ContentStart.GetCharacterRect(LogicalDirection.Forward);
```
Comment (jsuarezruiz) [inline src/Controls/src/Core/Platform/Windows/Extensions/FormattedStringExtensions.cs:149]: Always returning an empty Rect. Reviewing why is happening and possible workarounds.

**Comment by IeuanWalker** [comment]
@jsuarezruiz not sure if you have or havnt, but could you add a test for multiple clickable spans.

There was an issue in XF where if you had multiple spans with tap gestures, if you clicked on the last span, all the previous spans tap gestures would also trigger.

It was fixed, just want to make sure it doesnt crop back up in MAUI - https://github.com/xamarin/Xamarin.Forms/issues/11650

**Comment by jsuarezruiz** [comment]
> @jsuarezruiz not sure if you have or havnt, but could you add a test for multiple clickable spans.
> 
> There was an issue in XF where if you had multiple spans with tap gestures, if you clicked on the last span, all the previous spans tap gestures would also trigger.
> 
> It was fixed, just want to make sure it doesnt crop back up in MAUI - https://github.com/xamarin/Xamarin.Forms/issues/11650

Good point, I've checked it.
multiple-spans-click

**Comment by IeuanWalker** [comment]
@pictos also wondering if tap gestures will be updated in maui to support keyboard accessibility?

In XF navigating with a keyboard, tap gestured controls were skipped over.

**Comment by pictos** [comment]
> @pictos also wondering if tap gestures will be updated in maui to support keyboard accessibility?
> 
> 
> 
> In XF navigating with a keyboard, tap gestured controls were skipped over.

@PureWeen , I belive you have more context to answer this question

**Comment by SadamSkula** [comment]
Hi @jsuarezruiz thank you for hard work on implementing this one, as well as other things on MAUI. 
Do we have ETA for having this in production? 

Best regards,
Sadam

**Comment by hartez** [comment]
Pulling this out of draft, need someone to diagnose the failures from the Android API 23 tests.

File: src/Controls/samples/Controls.Sample/Pages/Controls/LabelPage.xaml.cs [hunk 2]
```diff
@@ -1,36 +1,71 @@
...
+
+		void SetRandomBackgroundColor(Span span)
+		{
+			var oldColor = span.BackgroundColor;
+
+			Color newColor;
+			do
+			{
+				newColor = _colors[_rand.Next(_colors.Length)];
+			} while (oldColor == newColor);
```
Comment (mandel-macaque) [inline src/Controls/samples/Controls.Sample/Pages/Controls/LabelPage.xaml.cs:66]: I see the use of the do while, but is it really needed? I mean, it just looks like you are doing:
```suggestion
			Color newColor = _colors[_rand.Next(_colors.Length)];
			while (oldColor == newColor) {
			    newColor = _colors[_rand.Next(_colors.Length)];
			}
```
I mention this because if color is a struct we should *be* ok, but if it is a ref type, the compiler is going to ask you to move it to Color? and then we have to use ? everywhere.

File: src/Controls/src/Core/HandlerImpl/Label/Label.Windows.cs [hunk 3]
```diff
@@ -1,15 +1,40 @@
...
 {
 	public partial class Label
 	{
+		TextBlock _textBlock;
+
+		private protected override void OnHandlerChangedCore()
+		{
+			base.OnHandlerChangedCore();
+
+			if (Handler != null)
```
Comment (mandel-macaque) [inline src/Controls/src/Core/HandlerImpl/Label/Label.Windows.cs:16]:
```suggestion
			if (Handler is not null)
```

File: src/Controls/src/Core/HandlerImpl/Label/Label.Windows.cs [hunk 4]
```diff
@@ -1,15 +1,40 @@
...
+			{
+				if (Handler is LabelHandler labelHandler && labelHandler.PlatformView is TextBlock textBlock)
+				{
+					_textBlock = textBlock;
+					_textBlock.SizeChanged += OnSizeChanged;
+				}
+			}
+			else
+			{
+				if (_textBlock != null)
```
Comment (mandel-macaque) [inline src/Controls/src/Core/HandlerImpl/Label/Label.Windows.cs:26]:
```suggestion
				if (_textBlock is not null)
```

File: src/Controls/src/Core/HandlerImpl/Label/Label.Windows.cs [hunk 5]
```diff
@@ -24,5 +49,56 @@ public partial class Label
...
 			handler.PlatformView?.UpdateMaxLines(label);
+
+		void OnSizeChanged(object sender, UI.Xaml.SizeChangedEventArgs e)
+		{
+			if (Handler is LabelHandler labelHandler)
+			{
+				var platformView = labelHandler.PlatformView;
+				var virtualView = labelHandler.VirtualView as Label;
+
+				if (platformView == null || virtualView == null)
```
Comment (mandel-macaque) [inline src/Controls/src/Core/HandlerImpl/Label/Label.Windows.cs:60]:
```suggestion
				if (platformView is null || virtualView is null)
```

File: src/Controls/src/Core/HandlerImpl/Label/Label.Windows.cs [hunk 6]
```diff
@@ -24,5 +49,56 @@ public partial class Label
...
+			IList<double> inlineHeights = new List<double>();
+
+			if (Handler is LabelHandler labelHandler)
+			{
+				var platformView = labelHandler.PlatformView;
+				var virtualView = labelHandler.VirtualView as Label;
+
+				FormattedString formatted = virtualView.FormattedText;
+
+				if (formatted != null)
```
Comment (mandel-macaque) [inline src/Controls/src/Core/HandlerImpl/Label/Label.Windows.cs:79]:
```suggestion
				if (formatted is not null)
```

File: src/Controls/src/Core/Platform/Android/Extensions/FormattedStringExtensions.cs [hunk 7]
```diff
@@ -137,27 +137,24 @@ public static void RecalculateSpanPositions(this TextView textView, Label elemen
 				return;
 
 			var labelWidth = finalSize.Request.Width;
-
```
Comment (mandel-macaque) [inline src/Controls/src/Core/Platform/Android/Extensions/FormattedStringExtensions.cs:140]: minimise the diff.

File: src/Controls/src/Core/Platform/Android/Extensions/FormattedStringExtensions.cs [hunk 8]
```diff
@@ -137,27 +137,24 @@ public static void RecalculateSpanPositions(this TextView textView, Label elemen
...
 			if (labelWidth <= 0 || finalSize.Request.Height <= 0)
 				return;
 
 			if (spannableString == null || spannableString.IsDisposed())
 				return;
 
 			var layout = textView.Layout;
-
```
Comment (mandel-macaque) [inline src/Controls/src/Core/Platform/Android/Extensions/FormattedStringExtensions.cs:148]: same

File: src/Controls/src/Core/Platform/Windows/Extensions/FormattedStringExtensions.cs [hunk 9]
```diff
@@ -127,5 +128,54 @@ public static void UpdateInlines(this TextBlock textBlock, Label label)
 
 			return Tuple.Create(run, span.TextColor, span.BackgroundColor);
 		}
+
+		internal static void RecalculateSpanPositions(this TextBlock control, Label element, IList<double> inlineHeights)
+		{
+			if (element?.FormattedText?.Spans == null
```
Comment (mandel-macaque) [inline src/Controls/src/Core/Platform/Windows/Extensions/FormattedStringExtensions.cs:134]:
```suggestion
			if (element?.FormattedText?.Spans is null	
```

File: src/Controls/src/Core/Platform/iOS/Extensions/FormattedStringExtensions.cs [hunk 10]
```diff
@@ -124,5 +127,129 @@ public static class FormattedStringExtensions
 
 			return attrString;
 		}
+
+		// TODO: Make public for .NET 8
+		internal static void RecalculateSpanPositions(this UILabel control, Label element)
+		{
+			if (element == null)
```
Comment (mandel-macaque) [inline src/Controls/src/Core/Platform/iOS/Extensions/FormattedStringExtensions.cs:134]:
```suggestion
			if (element is null)
```

File: src/Controls/src/Core/Platform/iOS/Extensions/FormattedStringExtensions.cs [hunk 11]
```diff
@@ -124,5 +127,129 @@ public static class FormattedStringExtensions
...
+				return;
+
+			if (element.TextType == TextType.Html)
+				return;
+
+			if (element?.FormattedText?.Spans == null
```
Comment (mandel-macaque) [inline src/Controls/src/Core/Platform/iOS/Extensions/FormattedStringExtensions.cs:140]:
```suggestion
			if (element?.FormattedText?.Spans is null
```

File: src/Controls/src/Core/Platform/iOS/Extensions/FormattedStringExtensions.cs [hunk 12]
```diff
@@ -124,5 +127,129 @@ public static class FormattedStringExtensions
...
+				return;
+
+			var finalSize = control.Frame;
+
+			if (finalSize.Width <= 0 || finalSize.Height <= 0)
+				return;
+
+			var inline = control.AttributedText;
+
+			if (inline == null)
```
Comment (mandel-macaque) [inline src/Controls/src/Core/Platform/iOS/Extensions/FormattedStringExtensions.cs:151]:
```suggestion
			if (inline is null)
```

File: src/Controls/src/Core/Platform/iOS/Extensions/FormattedStringExtensions.cs [hunk 13]
```diff
@@ -124,5 +127,129 @@ public static class FormattedStringExtensions
...
+				LineFragmentPadding = 0
+			};
+
+			layoutManager.AddTextContainer(textContainer);
+
+			var labelWidth = finalSize.Width;
+
+			var currentLocation = 0;
+
+			for (int i = 0; i < element.FormattedText.Spans.Count; i++)
```
Comment (mandel-macaque) [inline src/Controls/src/Core/Platform/iOS/Extensions/FormattedStringExtensions.cs:173]: I might be missing something, but is it 'i' use elsewhere in the loop, can we move to a foreach?

File: src/Controls/src/Core/Platform/iOS/Extensions/FormattedStringExtensions.cs [hunk 14]
```diff
@@ -124,5 +127,129 @@ public static class FormattedStringExtensions
...
+		}
+
+		static double FindDefaultLineHeight(this UILabel control, int start, int length)
+		{
+			if (length == 0)
+				return 0.0;
+
+			var textStorage = new NSTextStorage();
+#if __MOBILE__
+			if (control.AttributedText != null)
```
Comment (mandel-macaque) [inline src/Controls/src/Core/Platform/iOS/Extensions/FormattedStringExtensions.cs:236]:
```suggestion
			if (control.AttributedText is not null)
```

File: src/Controls/src/Core/Platform/iOS/Extensions/FormattedStringExtensions.cs [hunk 15]
```diff
@@ -124,5 +127,129 @@ public static class FormattedStringExtensions
...
+				textStorage.SetString(control.AttributedText.Substring(start, length));
+#else
+			if(control.AttributedStringValue != null)
```
Comment (mandel-macaque) [inline src/Controls/src/Core/Platform/iOS/Extensions/FormattedStringExtensions.cs:239]:
```suggestion
			if(control.AttributedStringValue is not null)
```

File: src/Core/tests/DeviceTests.Shared/Stubs/ContextStub.cs [hunk 16]
```diff
@@ -38,6 +38,12 @@ public object GetService(Type serviceType)
 
 			if (serviceType == typeof(NavigationRootManager))
 				return _windowManager ??= new NavigationRootManager(this);
+
+			if (serviceType == typeof(IFontManager))
```
Comment (mandel-macaque) [inline src/Core/tests/DeviceTests.Shared/Stubs/ContextStub.cs:42]: oh! do not need to do == against a the if you have the is operator
```suggestion
			if (serviceType is IFontManager)
```

**Review by mandel-macaque — CHANGES_REQUESTED** [review]

File: src/Core/tests/DeviceTests.Shared/Stubs/ContextStub.cs (hunk 16 above)
Comment (PureWeen) [inline src/Core/tests/DeviceTests.Shared/Stubs/ContextStub.cs:42]: @jsuarezruiz I think this change also is what broke the API 23 tests 

If I remove this code then the `FontFamilyInitializesCorrectly` tests all pass correctly.

Comment (PureWeen) [inline src/Core/tests/DeviceTests.Shared/Stubs/ContextStub.cs:42]: @jsuarezruiz I think this change also is what broke the API 23 tests 

If I remove this code then the `FontFamilyInitializesCorrectly` tests all pass correctly.

**Review by PureWeen — DISMISSED** [review]

File: src/Controls/src/Core/Platform/iOS/Extensions/FormattedStringExtensions.cs (hunk 13 above)
Comment (hartez) [reply src/Controls/src/Core/Platform/iOS/Extensions/FormattedStringExtensions.cs:173]: `Spans` is an `IList<Span>`, so `foreach` will allocate an enumerator. The `for` loop is a little less readable, but better for performance in this case.

File: src/Core/tests/DeviceTests.Shared/Stubs/ContextStub.cs (hunk 16 above)
Comment (mandel-macaque) [reply src/Core/tests/DeviceTests.Shared/Stubs/ContextStub.cs]: @jsuarezruiz don't want the suggestion?

File: src/Controls/samples/Controls.Sample/Pages/Controls/LabelPage.xaml.cs [hunk 17]
```diff
@@ -1,21 +1,32 @@
...
 			BindingContext = new LabelViewModel();
 		}
 
-		void ClickGestureRecognizer_Clicked(object sender, System.EventArgs e)
+		void ClickGestureRecognizer_Clicked(System.Object sender, System.EventArgs e)
 		{
+			var label = sender as Label;
+
+			if (label == null)
+				return;
```
Comment (mandel-macaque) [inline src/Controls/samples/Controls.Sample/Pages/Controls/LabelPage.xaml.cs:26]:
```suggestion
			if (sender is not Label)
				return;
```

**Review by mandel-macaque — DISMISSED** [review]

File: src/Controls/src/Core/Platform/GestureManager/GesturePlatformManager.Android.cs [hunk 18]
```diff
@@ -176,10 +176,11 @@ void SetupGestures()
 				return;
 
 			var platformView = Control;
+
 			if (platformView == null)
 				return;
 
-			if (View.GestureRecognizers.Count == 0)
+			if (View.GestureController.CompositeGestureRecognizers.Count == 0)
```
Comment (rmarinho) [inline src/Controls/src/Core/Platform/GestureManager/GesturePlatformManager.Android.cs:183]: Why this change?

File: src/Controls/src/Core/HandlerImpl/Label/Label.Windows.cs [hunk 19]
```diff
@@ -1,15 +1,40 @@
...
+
+		private protected override void OnHandlerChangedCore()
+		{
+			base.OnHandlerChangedCore();
+
+			if (Handler is not null)
+			{
+				if (Handler is LabelHandler labelHandler && labelHandler.PlatformView is TextBlock textBlock)
+				{
+					_textBlock = textBlock;
```
Comment (rmarinho) [inline src/Controls/src/Core/HandlerImpl/Label/Label.Windows.cs:20]: Why do we need the native view change here? can't we look at size changes on the Label itself ? Override Measure or something ?

File: src/Controls/src/Core/Platform/Windows/Extensions/FormattedStringExtensions.cs [hunk 20]
```diff
@@ -127,5 +128,54 @@ public static void UpdateInlines(this TextBlock textBlock, Label label)
 
 			return Tuple.Create(run, span.TextColor, span.BackgroundColor);
 		}
+
+		internal static void RecalculateSpanPositions(this TextBlock control, Label element, IList<double> inlineHeights)
```
Comment (rmarinho) [inline src/Controls/src/Core/Platform/Windows/Extensions/FormattedStringExtensions.cs:132]: if we make the other public we should do it for this as well

Comment (jsuarezruiz) [reply src/Controls/src/Core/Platform/Windows/Extensions/FormattedStringExtensions.cs]: Changed.

File: src/Controls/src/Core/HandlerImpl/Label/Label.Windows.cs [hunk 21]
```diff
@@ -24,5 +49,56 @@ public partial class Label
...
+			if (Handler is LabelHandler labelHandler)
+			{
+				var platformView = labelHandler.PlatformView;
+				var virtualView = labelHandler.VirtualView as Label;
+
+				FormattedString formatted = virtualView.FormattedText;
+
+				if (formatted is not null)
+				{
+					var fontManager = virtualView.RequireFontManager();
```
Comment (mandel-macaque) [inline src/Controls/src/Core/HandlerImpl/Label/Label.Windows.cs:81]: virtualView can be null because we used an as casting in line 75. This could throw a NRE.

File: src/Controls/src/Core/HandlerImpl/Label/Label.Windows.cs [hunk 22]
```diff
@@ -24,5 +49,56 @@ public partial class Label
...
+					platformView.Inlines.Clear();
+
+					// Have to implement a measure here, otherwise inline.ContentStart and ContentEnd will be null, when used in RecalculatePositions
+					platformView.Measure(new global::Windows.Foundation.Size(double.MaxValue, double.MaxValue));
+
+					var heights = new List<double>();
+					for (var i = 0; i < formatted.Spans.Count; i++)
+					{
+						var span = formatted.Spans[i];
```
[the remark this answers is not in the document]
Comment (mandel-macaque) [reply src/Controls/src/Core/HandlerImpl/Label/Label.Windows.cs:90]: @jsuarezruiz  ?

File: src/Controls/src/Core/Platform/Windows/Extensions/FormattedStringExtensions.cs (hunk 9 above)
Comment (mandel-macaque) [reply src/Controls/src/Core/Platform/Windows/Extensions/FormattedStringExtensions.cs:134]: Still applies.

**Review by mandel-macaque — CHANGES_REQUESTED** [review]
Some comments need to still be addressed and we have some NRE cases in the code.

File: src/Controls/src/Core/HandlerImpl/Label/Label.Windows.cs (hunk 22 above)
[the remark this answers is not in the document]
Comment (mandel-macaque) [reply src/Controls/src/Core/HandlerImpl/Label/Label.Windows.cs:90]: Still stands.

**Review by mandel-macaque — CHANGES_REQUESTED** [review]

**Review by hartez — CHANGES_REQUESTED** [review]
The targets seem a bit off on iOS.

Testing this on iOS, I can tap above and below the "Click Me" Label and it changes the colors - for example, I can tap the "Colors" Label above "Click Me" and the event fires.

Same thing with the part that says "Two clickable spans in one line" - I can tap the word "line" and it causes the color change.

File: src/Controls/src/Core/HandlerImpl/Label/Label.Windows.cs (hunk 21 above)
Comment (jsuarezruiz) [reply src/Controls/src/Core/HandlerImpl/Label/Label.Windows.cs]: Right, I have applied changes to avoid possible NRE.

**Comment by jsuarezruiz** [comment]
This PR affects all platforms, and has quite a few changes here and there. To facilitate its review and have fixes in releases as soon as possible, I am going to close this PR to open smaller PRs focused on each platform.
